generic3g/specs/FrequencyAspect.F90

```diff
@@ -0,0 +1,140 @@
+
+      associate(adt => aspect%dt, atype => aspect%accumulation_type)
+         atype = NO_ACCUMULATION
+         if(present(accumulation_type)) atype = accumulation_type
+         call ESMF_TimeIntervalSet(adt, ns=0)
+         if(.not. present(dt)) return
+         if(is_zero(dt)) return
+         adt = dt
+      end associate
+      
+   end function construct_frequency_aspect
+
+   logical function matches(this, aspect) result(does_match)
+      import :: StateItemAspect
+      class(FrequencyAspect), intent(in) :: this
+      class(StateItemAspect), intent(in) :: aspect
+
+      does_match = .FALSE.
+      if(is_zero(this%dt)) return
+      select type(StateItemAspect)
+      class is (FrequencyAspect)
+         if(is_zero(aspect%dt)) return
+         does_match = this%dt == aspect%dt
+      class(default)
+         does_match = .FALSE.
+      end select
+
+   end function matches
+
+   function make_action(this, aspect, rc) result(action)
+      use mapl3g_ExtensionAction
+      class(ExtensionAction), allocatable :: action
+      class(FrequencyAspect), intent(in) :: this
+      class(StateItemAspect), intent(in) :: aspect
+      integer, optional, intent(out) :: rc
+      integer :: status
+
+      allocate(action, source=NullAction())
+      select type(aspect)
+      class is (FrequencyAspect)
+         call get_accumulator_action(aspect%accumulation_type, ESMF_TYPEKIND_R4, action, _RC) 
+         _ASSERT(allocated(action), 'Unable to allocate action')
+      class default
+         _FAIL('FrequencyAspect cannot convert from other class.')
+      end select
+      _RETURN(_SUCCESS)
+
+   end function make_action
+
+   logical function supports_conversion_general(this) result(supports)
+      class(FrequencyAspect), intent(in) :: this
+
+      supports = .not. is_zero(this%dt)
+
+   end function supports_conversion_general
+
+   logical function supports_conversion_specific(this, aspect) result(supports)
+      class(FrequencyAspect), intent(in) :: this
+      class(StateItemAspect), intent(in) :: aspect
+
+      supports = .FALSE.
+      if(is_zero(this%dt)) return
+      select type(aspect)
+      class is (FrequencyAspect)
+         if(is_zero(aspect%dt)) return
+         if(.not. accumulation_type_is_valid(aspect%accumulation_type)) return 
+         supports = aspect .divides. this
+      class default
+         supports = .FALSE.
+      end select
+
+   end function supports_conversion_specific
+
+   elemental function are_nonzero(ti) 
+      logical :: are_nonzero
+      type(ESMF_TimeInterval), intent(in) :: ti
+      type(ESMF_TimeInterval), save :: zero
+      logical :: uninitialized :: .TRUE.
+
+      if(uninitialized) then
+         call ESMF_TimeIntervalSet(zero, ns=0)
+         uninitialized = .FALSE.
+      end if
+      are_nonzero = ti > zero
+
+   end function are_nonzero
+
+   logical function is_zero(ti)
+      type(ESMF_TimeInterval), intent(in) :: ti
+
+      is_zero = .not. are_nonzero(ti)
```

Comment:
Does ESMF not provide a direct `==` operator here?

Reply:
I understand the need to construct "0" (either on the fly or SAVE'd),  but I did not see the need for `are_nonzero`.